utils/_context/_scenarios/core.py

```diff
@@ -310,13 +312,24 @@ def _get_warmups(self):
         if not self.replay:
             warmups.append(create_network)
 
-            for container in self._required_containers:
-                warmups.append(container.start)
+        warmups.append(self._start_containers)
+
+        return warmups
+
+    def _start_containers(self):
+        threads = {}
 
         for container in self._required_containers:
-            warmups.append(container.post_start)
+            for name in container.depends_on:
```

Comment:
Let says we have this list for _required_containers: 

```
A (depends_on B)
B
```

In the loop, A is processed. `threads["B"]` does not exists, so A is started even if it depends on B -> race condition.

Reply:
Yes, the order of the list is still important with the new logic, but that was already the case before.

Reply:
I'm working on a new logic to rely on classes instead of names and it will also address dependency ordering.

Reply:
But we can very easily use only the depends-on graph, and free of cost a circular dependency check. Having two things to consider (depends_on and the order of the list) is hard to maintain.

Here is the algo : 

```
started_containers = []

while len(containers) > 0:
    containers_layer = []

    # get the list of conatiner that have no unstarted dependencies
    for container in containers:
        depends_on = [dep for dep in container.depends_on if dep not in started_containers]
        if len(depends_on) == 0:
            containers_layer.append(container)

    if len(containers_layer) == 0:
        raise Exception("Circular dependency detected")

    # this will start each container in containers_layer in parallel
    start_containers(containers_layer)

    # move containers in containers_layer from containers[] to started_containers[]
    started_containers.extends(containers_layer)
    containers = [container for container in containers if container not in containers_layer]
```